improver/generate_ancillaries/generate_timezone_mask.py

```diff
@@ -0,0 +1,351 @@
+        """
+        Configure plugin options to generate the desired ancillary.
+
+        Args:
+            ignore_dst (Optional[bool]):
+                If True, find and use the UTC offset to a grid point ignoring
+                daylights savings.
+            time (Optional[str]):
+                A datetime specified in the format YYYYMMDDTHHMMZ at which to
+                calculate the mask (UTC). If daylights savings are ignored this
+                will have no impact on the resulting masks.
+            groupings (Optional[Dict[int, List[int, int]]]):
+                A dictionary specifying how timezones should be grouped if so
+                desired. This dictionary takes the form::
+
+                    {-9: [-12, -5], 0:[-4, 4], 9: [5, 12]}
+
+                The numbers in the lists denote the inclusive limits of the
+                groups. This is of use if data is not available at hourly
+                intervals.
+        """
+        from timezonefinder import TimezoneFinder
+
+        self.tf = TimezoneFinder()
+        self.time = time
+        self.ignore_dst = ignore_dst
+        self.groupings = groupings
+
+    def _set_time(self, cube):
+        """
+        Set self.time to a datetime object specifying the date and time for
+        which the masks should be created. self.time is set in UTC.
+
+        Args:
+            cube (iris.cube.Cube):
+                The cube from which the validity time should be taken is one
+                has not been explicitly provided by the user.
+        """
+        if self.time:
+            self.time = datetime.strptime(self.time, "%Y%m%dT%H%MZ")
+            self.time = pytz.utc.localize(self.time)
+        else:
+            try:
+                self.time = cube.coord("time").cell(0).point
+                self.time = pytz.utc.localize(self.time)
+            except CoordinateNotFoundError:
+                msg = (
+                    "The input cube does not contain a 'time' coordinate. "
+                    "As such a time must be provided by the user."
+                )
+                raise ValueError(msg)
+
+    @staticmethod
+    def _get_coordinate_pairs(cube):
+        """
+        Create an array containing all the pairs of coordinates that describe
+        y-x points in the grid.
+
+        Args:
+            cube (iris.cube.Cube):
+                The cube from which the y-x grid is being taken.
+        Returns:
+            numpy.array:
+                A numpy array containing all the pairs of coordinates that describe
+                the y-x points in the grid. This array is 2-dimensional, being
+                2 by the product of the grid's y-x dimension lengths.
+        """
+        if lat_lon_determine(cube) is not None:
+            yy, xx = transform_grid_to_lat_lon(cube)
+        else:
+            latitudes = cube.coord("latitude").points
+            longitudes = cube.coord("longitude").points.copy()
```

Comment:
Yes, because the longitudes are modified below if there are any >180 degrees. I don't want to modify these in place in the cube, so I have copied them. The latitudes are not modified, so I don't need to.